quic/s2n-quic-platform/src/socket/ring.rs

```diff
@@ -0,0 +1,307 @@
+        let ptr = self.cursor.data_ptr();
+        unsafe {
+            let ptr = ptr.as_ptr().add(idx as _);
+            core::slice::from_raw_parts_mut(ptr, len as _)
+        }
+    }
+
+    /// Returns true if the producer is not closed
+    #[inline]
+    pub fn is_open(&self) -> bool {
+        self.wakers.is_open()
+    }
+}
+
+/// A producer ring for messages
+pub struct Producer<T: Message> {
+    cursor: Cursor<T>,
+    wakers: atomic_waker::Handle,
+    #[allow(dead_code)]
+    storage: Arc<message::Storage>,
+}
+
+/// Safety: Storage is synchronized with the Cursor
+unsafe impl<T: Message> Send for Producer<T> {}
+/// Safety: Storage is synchronized with the Cursor
+unsafe impl<T: Message> Sync for Producer<T> {}
+
+impl<T: Message> Producer<T> {
+    /// Acquires capacity for sending messages to the consumer
+    #[inline]
+    pub fn acquire(&mut self, watermark: u32) -> u32 {
+        self.cursor.acquire_producer(watermark)
+    }
+
+    /// Polls capacity for sending messages to the consumer
+    #[inline]
+    pub fn poll_acquire(&mut self, watermark: u32, cx: &mut Context) -> Poll<u32> {
+        macro_rules! try_acquire {
+            () => {{
+                let count = self.acquire(watermark);
+
+                if count > 0 {
+                    return Poll::Ready(count);
+                }
+            }};
+        }
+
+        try_acquire!();
+
+        self.wakers.register(cx.waker());
+
+        try_acquire!();
+
+        Poll::Pending
+    }
+
+    /// Releases ready-to-consume messages to the consumer
+    #[inline]
+    pub fn release(&mut self, len: u32) {
+        if len == 0 {
+            return;
+        }
+
+        debug_assert!(len <= self.cursor.cached_producer_len());
+
+        let idx = self.cursor.cached_producer();
+        let size = self.cursor.capacity();
+
+        // replicate any written items to the secondary region
+        unsafe {
+            let replication_count = (size - idx).min(len);
+
+            debug_assert_ne!(replication_count, 0);
+
+            let ptr = self.cursor.data_ptr().as_ptr().add(idx as _);
+
+            let primary = ptr;
+            let secondary = ptr.add(size as _);
+
+            self.replicate(primary, secondary, replication_count as _);
+        }
+
+        // if messages were also written to the secondary region, we need to copy them back to the
+        // primary region
+        if let Some(replication_count) = (idx + len).checked_sub(size).filter(|v| *v > 0) {
+            unsafe {
+                let ptr = self.cursor.data_ptr().as_ptr();
+
+                let primary = ptr;
+                let secondary = ptr.add(size as _);
+
+                self.replicate(secondary, primary, replication_count as _);
+            }
+        }
+
+        self.cursor.release_producer(len);
+
+        self.wakers.wake();
+    }
+
+    /// Returns the empty messages for the producer
+    #[inline]
+    pub fn data(&mut self) -> &mut [T] {
```

Comment:
I guess we already asserted that T here is initialized? Technically this is more of `&mut [MaybeUninit<T>]` probably in terms of how the caller *should* treat it -- I wonder if that adds enough verbosity that it's not worth making that the return type. It would help catch typos where we start trying to read from the array while being layout-equivalent.

Reply:
By the time you're calling `data`, `T` will be initialized, yes. It's kind of a weird one cause we have initialized the message with valid pointers and lengths, but the actual payload data is garbage.